src/clusterfuzz/_internal/system/archive.py

```diff
@@ -34,111 +36,316 @@
-
-      for info in tar_file.getmembers():
-        if not file_match_callback(info.name):
-          continue
-
-        yield ArchiveFile(info.name, info.size,
-                          maybe_extract(tar_file.extractfile, info))
-      tar_file.close()
-    except tarfile.TarError:
-      logs.log_error('Bad tar file %s.' % archive_path)
-
-  elif archive_type == ArchiveType.TAR_LZMA:
-    assert archive_obj is None, "LZMAFile doesn't support opening file handles."
+      return self.open(member)
+    except:
+      return None
+
+  def get_first_file_matching(self, search_string):
+    """Gets the name of the first matching member.
+
+    Args:
+        search_string (str): the string to be searched for.
+
+    Returns:
+        str: the member name that matched.
+    """
+    for file in self.list_files():
+      if file.filename.startswith('__MACOSX/'):
+        # Exclude MAC resource forks.
+        continue
+
+      if search_string in file.filename:
+        return file.filename
+    return None
+
+  def extracted_size(self, file_match_callback=None):
+    """Gets the total extracted size of the file matched by
+    file_match_callback. If file_match_callback is None, gets the extracted
+    size of the whole archive.
+
+    Args:
+        file_match_callback (optional): the file matching callback. Defaults
+        to None.
+
+    Returns:
+        int: the extracted size.
+    """
+    return sum(f.file_size_bytes
+               for f in self.list_files()
+               if not file_match_callback or file_match_callback(f.filename))
+
+
+class TarArchiveReader(ArchiveReader):
+  """A tar archive reader. Currently supports TAR and TAR_LZMA archives.
+  """
+
+  def __init__(self, archive_path, file_obj=None) -> None:
+    if file_obj:
+      self._archive = tarfile.open(fileobj=file_obj)
+    else:
+      archive_type = get_archive_type(archive_path)
+      mode = 'r' if archive_type == ArchiveType.TAR else 'r:xz'
+      self._archive = tarfile.open(archive_path, mode=mode)
+    self._archive_path = archive_path
+
+  def list_files(self) -> typing.List[ArchiveMemberInfo]:
+    return [
+        ArchiveMemberInfo(
+            filename=f.name,
+            is_dir=f.isdir(),
+            file_size_bytes=f.size,
+            mode=f.mode) for f in self._archive.getmembers()
+    ]
+
+  def open(self, member):
+    return self._archive.extractfile(member)
+
+  def close(self) -> None:
+    self._archive.close()
+
+  def extract(self, member, path, trusted=False):
+    # If the output directory is a symlink, get its actual path since we will be
+    # doing directory traversal checks later when unpacking the archive.
+    output_directory = os.path.realpath(path)
+
+    if not trusted:
+      if (_is_attempting_path_traversal(self._archive_path, output_directory,
+                                        member)):
+        return None
+
+    self._archive.extract(member=member, path=output_directory)
```

Comment:
Consider extracting (no pun intended) the path traversal check to `ArchiveReader`:

```py
def extract(...):
  if attempting_traversal(...):
    return None

  return self._extract(...)

@abc.abstractmethod
def _extract(...):
  raise NotImplementedError
```